lib/elixir/lib/system.ex

```diff
@@ -409,15 +425,127 @@ defmodule System do
 
-  If the VM terminates programmatically, via `System.stop/1` or `System.halt/1`,
-  the `at_exit/1` callbacks are not executed.
+  If the VM terminates programmatically, via `System.stop/1`, `System.halt/1`,
+  or exit signals, the `at_exit/1` callbacks are not executed.
   """
   @spec at_exit((non_neg_integer -> any)) :: :ok
   def at_exit(fun) when is_function(fun, 1) do
     :elixir_config.update(:at_exit, &[fun | &1])
     :ok
   end
 
+  defmodule SignalHandler do
+    @moduledoc false
+    @behaviour :gen_event
+
+    @impl true
+    def init({event, fun}) do
+      {:ok, {event, fun}}
+    end
+
+    @impl true
+    def handle_call(_message, state) do
+      {:ok, :ok, state}
+    end
+
+    @impl true
+    def handle_event(signal, {event, fun}) do
+      if signal == event, do: :ok = fun.()
+      {:ok, {event, fun}}
+    end
+
+    @impl true
+    def handle_info(_, {event, fun}) do
+      {:ok, {event, fun}}
+    end
+  end
+
+  @doc """
+  Traps the given `signal` to execute the `fun`.
+
+  An optional `id` that uniquely identifies the function
+  can be given, otherwise a unique one is automatically
+  generated. If a previously registered `id` is given,
+  this function returns an error tuple. The `id` can be
+  used to remove a registered signal by calling
+  `untrap_signal/1`.
+
+  The given `fun` receives no arguments and it must return
+  `:ok`.
+
+  The first time a signal is trapped, it will override the
+  default behaviour from the operating system. If the same
+  signal is trapped multiple times, subsequent functions
+  given to `trap_signal` will execute *first*. In other
+  words, you can consider each function is prepended to
+  the signal handler.
+
+  By default, the Erlang VM register traps to the three
+  signals:
+
+    * `:sigstop` - gracefully shuts down the VM with `stop/0`
+    * `:sigquit` - halts the VM via `halt/0`
+    * `:sigusr1` - halts the VM via status code of 1
+
+  Therefore, if you add traps to the signals above, the
+  default behaviour above will be executed after all user
+  signals.
+
+  **Important**: all signals run from a single process,
+  therefore, blocking the `fun` will block subsequent traps.
+  It is also not possible to add or remove traps from within
+  a trap itself.
+  """
+  @spec trap_signal(signal, (() -> :ok)) :: {:ok, reference()}
+  @spec trap_signal(signal, id, (() -> :ok)) :: {:ok, id} | {:error, :already_registered}
+        when id: term()
+  def trap_signal(signal, id \\ make_ref(), handler)
+      when signal in @signals and is_function(handler, 0) do
+    :elixir_config.serial(fn ->
+      gen_id = {signal, id}
+
+      if {SignalHandler, gen_id} in signal_handlers() do
+        {:error, :already_registered}
+      else
+        :ok =
+          :gen_event.add_handler(:erl_signal_server, {SignalHandler, gen_id}, {signal, handler})
+
+        :os.set_signal(signal, :handle)
+        {:ok, id}
+      end
+    end)
+  end
```

Comment:
To match the documentation:

```suggestion
  def trap_signal(signal, id \\ make_ref(), fun)
      when signal in @signals and is_function(fun, 0) do
    :elixir_config.serial(fn ->
      gen_id = {signal, id}

      if {SignalHandler, gen_id} in signal_handlers() do
        {:error, :already_registered}
      else
        :ok =
          :gen_event.add_handler(:erl_signal_server, {SignalHandler, gen_id}, {signal, fun})

        :os.set_signal(signal, :handle)
        {:ok, id}
      end
    end)
  end
```